prov/verbs/src/verbs_info.c

```diff
@@ -726,6 +730,72 @@ static int vrb_have_device(void)
 	return ret;
 }
 
+static void vrb_disable_hmem(struct fi_info *info)
+{
+	info->caps &= ~FI_HMEM;
+	info->tx_attr->caps &= ~FI_HMEM;
+	info->rx_attr->caps &= ~FI_HMEM;
+	info->domain_attr->mr_mode &= ~FI_MR_HMEM;
+}
+
+static void vrb_alter_hmem_attrs(struct fi_info *info)
+{
+	static const char *hmem_devices[] = {
+		"mlx4",
+		"mlx5",
+	};
+	static const char *kallsyms = "/proc/kallsyms";
+	static const char *hmem_symbol = "ib_register_peer_memory_client";
+	char *line = NULL;
+	size_t line_size = 0;
+	ssize_t bytes;
+	FILE *kallsyms_fd;
+	bool hmem_symbol_exists = false;
+	int i;
+
+	kallsyms_fd = fopen("/proc/kallsyms", "r");
+	if (!kallsyms_fd) {
+		vrb_disable_hmem(info);
+
+		VERBS_INFO(FI_LOG_CORE,
+			   "Failed to open %s: errno=%d. No HMEM support\n",
+			   kallsyms, -errno);
+		return;
+	}
+
+	while ((bytes = getline(&line, &line_size, kallsyms_fd)) != -1) {
```

Comment:
Is this the only way to determine if the kernel driver supports this?  Is there no capability bit or something?

Reply:
I have not seen a capability bit saying IB peer memory is supported or not. I can rework this to only scan the kernel symbols once.